### ISSUE TITLE
Feature request: ODBC reader / serializer (BrainScript)

### ISSUE THREAD

**Issue by ghost** [Microsoft/CNTK#1289]
Covering use cases where training data resides in DBs would differentiate and increase CNTK appeal as a platform.

**Comment by sayanpa** [comment]
Thanks for the suggestion. In our feature backlog now.

**Comment by wolfma61** [comment]
we spend a lot of effort in optimizing reader performance. A database isn't the perfect place for random read access, minibatch creation etc. for your training data, and you will have to guarantee repeatable/read semantic for your supplied dataset, not something your database admin is going to be very excited about if you have a long running training session

In the short term, copying your database data to an already supported dataformat is the easy and efficient workaround. 

Wolfgang

**Comment by avader906** [comment]
My initial feedback, working with TB distributed sets, is that CNTK is not capable to meet our demands. This is the context of all my feedback - none of it is personal. The effort you and the team deploys is tremendous. The main concern is to direct it to places CNTK adds value.

@wolfma61 I suggest you justify your statements. "Your database admin will not be happy, we know it best" is not acceptable. It shows two things
- poor understanding of business needs nor the type of infrastructure HPC users run. This is unusual, given elsewhere inside MSFT, initiatives to merge DNN type learning with big data are ongoing - sql server / R studio type integration for instance.
- the project needs to plug the lack of knowledge that leads to statements like "A database isn't the perfect place for random read access"


In other words, there should be none of that arrogance here,

Currently there is no capability or clear roadmap for efficiently handling TB type learning sets. The flat file reading is not adapted. The design is not efficient. Require the user to specify features and labels aliases - two chars minimum per sample is anything but (to note one obvious ugliness).

Once you start to adapt and optimize the readers to avoid parsing entire files, once you try to parallelarize the reading in the most optimal ways, specifying input masks, avoiding reading-through entire input and moving to binary you will recreate just another loading framework.

I can assure you that the effort you will be forced to spend there - hour for hour - is magnitudes more the effort to clean-up the reader API and bridge it to ODBC/OLDEDB or native (even if SQL Server only).

In fact, minibatch creation should be done in the data tier. All modern data tiers have efficient ways of sampling or sub-sampling. And even if that's too broad to address, all you need is dataset row index. Ideally, specifying the list of fields and an SQL statement to pull them should be everything CNTK needs, once it has the row index. 

CNTK added value is tiny - considering other more efficient ways to distribute and work with data. If you want the support of HPC public, you should pay attention to what we are telling you.  Issues like multiplying 100 columns on TB scale is not trivial. And the times when the host OS provided the most optimal I/O are long gone.

**Comment by wolfma61** [comment]
I think I agree with this analysis ... an tight integration with the database landscape is necessary, I just don't think we should hide this behind 'you need to talk to odbc/oldebb', much more is required to make efficient use of the db-backend, or let the server make efficient use of a ML component. for example, in a large (distributed) database and a multi-node ml-system ... who should make the decision of how data (minibatches) are distributed? I can see that a lot of these decisions and other can (should?)happen in the db-engine.
so I believe that just oledb/odbc isn't what you want, you want to tight orchestration between the db and ml landscape

so I might have given the impression that I believe that flat files are the answer to this problem, but I agree with you that this is not the solution, I want much better access to distributed data, and that requires smart integrataion/cooperation between these domains.

as you probably know SQLServer is already having a tight ML integration, and we are actively talking to that team about more integration with CNTK, but I can't say more than that (as it would be speculation on my part).

**Comment by avader906** [comment]
SQL Server integration is feasible if the code base is structured around stateless logic. In fact it would be trivial to market CNTK along SQL Server other Entreprise features. One of the easier ways is to to drop mpi, select the leaner of DCOM / ORPC and make sure no issues netdirect/RDMA/RoCE. Middle tier compute GPU nodes, data fed fetched by middle tier through SQL server based on parametrization and gradients propagation on the master node(s). But this is a fork.

More issues to come if the framework starts to turn client/server, some of them could be anticipated if reader team spends time on declarative API. Including ODBC reader feature would invite moving away from concepts such as minibatches. Nodes should be indifferent to where and how gradient aggregation happens. Some of this is already happening if you look at ASGD.

**Comment by raaaar** [comment]
We are working on simplifying the deserializer interface, which will be limited strictly to reading chunks of data, with the randomization and packing handled by the up-stack components. Once this is done (and documented), it should be fairly straight-forward to implement an ODBC deserializer and we'll be happy to accept it as a contribution. 

Adding @eldakms  to update this thread as soon as the deserilizer interface is published.

**Comment by eldakms** [comment]
Please have a look at Manual/How to write custom deserializer in python. The interface should simplify implementation of your own deserializers.
